x/programs/v2/program.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package v2
+
+import "C"
+import (
+	"context"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/hypersdk/state"
+	"github.com/bytecodealliance/wasmtime-go/v14"
+	"github.com/near/borsh-go"
+)
+
+const (
+	AllocName  = "alloc"
+	MemoryName = "memory"
+)
+
+type Context struct {
+	ProgramID ids.ID `json:"program"`
+	// Actor            [32]byte `json:"actor"`
+	// OriginatingActor [32]byte `json:"originating_actor"`
+}
+
+type CallInfo struct {
+	State        state.Mutable
+	Actor        ids.ID
+	Account      ids.ID
+	ProgramID    ids.ID
+	Fuel         uint64
+	FunctionName string
+	Params       []byte
+	inst         *ProgramInstance
+}
+
+func (c *CallInfo) RemainingFuel() uint64 {
+	remaining := c.Fuel
+	usedFuel, fuelEnabled := c.inst.store.FuelConsumed()
+	if fuelEnabled {
+		remaining -= usedFuel
+	}
+	return remaining
+}
+
+func (c *CallInfo) ConsumeFuel(fuel uint64) {
+	c.inst.store.ConsumeFuel(fuel)
+}
+
+type Program struct {
+	module    *wasmtime.Module
+	programID ids.ID
+}
+
+type ProgramInstance struct {
+	*Program
```

Comment:
What's going on here... is this a nameless property?

Reply:
yes/no.  I don't recall why that was on there in the first place, so I'll delete it.  What it is used for is embedding the struct into another, while preserving the ability to treat the enclosing struct as if it were of the same type as the embedded struct.  Its mostly syntactic sugar so you can call the fields and methods on the bigger struct instead of having to do inst.program.method.

Reply:
https://gobyexample.com/struct-embedding